lib/shopify_api/api_version.rb

```diff
@@ -3,111 +3,143 @@ module ShopifyAPI
   class ApiVersion
     class UnknownVersion < StandardError; end
     class InvalidVersion < StandardError; end
+    include Comparable
 
-    extend DefinedVersions
+    HANDLE_FORMAT = /^\d{4}-\d{2}$/.freeze
+    UNSTABLE_HANDLE = 'unstable'
+    UNSTABLE_AS_DATE = Time.utc(3000, 1, 1)
+    API_PREFIX = '/admin/api/'
+    COERSION_MODES = [:predefined_only, :define_on_unknown].freeze
 
-    include Comparable
+    class << self
+      attr_reader :coercion_mode
 
-    def self.coerce_to_version(version_or_name)
-      return version_or_name if version_or_name.is_a?(ApiVersion)
+      def versions
+        @versions ||= @coercer.known_versions
+      end
 
-      @versions ||= {}
-      @versions.fetch(version_or_name.to_s) do
-        raise UnknownVersion, "#{version_or_name} is not in the defined version set: #{@versions.keys.join(', ')}"
+      def coercer
+        @coercer ||= VersionCoercers::GenerateVersion.new
       end
-    end
 
-    def self.define_version(version)
-      @versions ||= {}
+      def coercion_mode=(mode)
```

Comment:
assignment of the coercer replaced the symbols it also allows for extension by implementing any coercer you want.

Reply:
ah yes, sorry. That got lost in translation last night. I'm not sure what else someone would want to do though beyond always creating a new version and raising an error if a version is not in the known set. If we had another use case I think I'd be more open to using a coercer class like this. 

Do you have an idea of any future needs that would require this extra flexibility and abstraction from the ApiVersion class?